### PR TITLE
LRs updates are called at the end of a skipped epoch

### LINKED ISSUE #21296: LR is not updated when `on_train_batch_start` returns -1
### Bug description

# Context

I'm training in a DDP setting on IterableDataset instances that can return different number of batches. To avoid any deadlock issue due to one GPU receiving more batches than others, I use a dataloader that yields a signal when it has exhausted the original iterator. I use this signal to stop the training loop by returning -1 in the `on_train_batch_start` hook.

```python
def on_train_batch_start(self, batch: Batch | ExhaustedDataloaderSignal, batch_idx: int):
        # Check if all dataloaders have data
        # If not, stop training for current epoch to avoid deadlock issues between GPUs
        has_data = True
        if isinstance(batch, ExhaustedDataloaderSignal):
            logger.debug(f"Rank {self.local_rank} has exhausted data on batch {batch_idx}.")
            has_data = False
        if self.trainer.world_size > 1:
            # Convert has_data as tensor for synchronization
            has_data = torch.tensor(has_data, dtype=torch.bool)
            # Synchronize has_data across ranks
            has_data = self.all_gather(has_data)
            # Check if all ranks have data
            has_data = bool(has_data.all().cpu())
        if not has_data:
            # Stop training for current epoch
            return -1
       ...
```
 
Each training epoch finishes as expected, except **the learning rate is not updated** according to the scheduler that has been specified.

# Expected behaviour

The epoch terminates and the learning rate are updated.

# Related documentation

[The documentation for the `on_train_batch_start`](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#on-train-batch-start) reads:

> If you return -1 here, you will skip training for the rest of the current epoch.

It is not clear if this implies that learning rate scheduler updates occurring at the end of the epoch will be also skipped.

# Likely origin of the problem

While checking [the code](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/loops/training_epoch_loop.py#L330-L359), one can see that if the hook returns `-1`, a `StopIteration` is raised, and the learning rate scheduler updates are never called. 

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

Define a lightning module with a learning rate scheduler with 1 epoch frequency and `on_train_batch_start` returning -1 during an epoch.

### Error messages and logs

No error message is produced. LR simply doesn't get updated.

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0): 2.5.5
#- PyTorch Version (e.g., 2.5): 2.8.0+cu126
#- Python version (e.g., 3.12): 3.13
#- OS (e.g., Linux): Linux
#- How you installed Lightning(`conda`, `pip`, source): uv pip
```

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by LTMeyer**
## What does this PR do?

This PR fixes learning rate not being updated at the end of epoch when `on_train_batch_start` returns -1.

Fixes #21296

It postpones the existing `raise StopIteration` after the learning rate update.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs): **Related issue #21296 hasn't been discussed yet.**
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section? **Yes**
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary) **No**
- [x] Did you write any **new necessary tests**? (not for typos and docs) **Yes**
- [x] Did you verify new and **existing tests pass** locally with your changes? **Yes**
- [x] Did you list all the **breaking changes** introduced by this pull request? **None**
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors) **Yes**

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21307.org.readthedocs.build/en/21307/

### CODE REVIEW

File: src/lightning/fabric/CHANGELOG.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
--
+- Learning rate scheduler is stepped at the end of epoch when `on_train_batch_start` returns -1 ([[#21296](https://github.com/Lightning-AI/pytorch-lightning/issues/21296)]).
```
Comment (SkafteNicki) [inline src/lightning/fabric/CHANGELOG.md]:
```suggestion
- Learning rate scheduler is stepped at the end of epoch when `on_train_batch_start` returns -1 ([#21296](https://github.com/Lightning-AI/pytorch-lightning/issues/21296)).
```

File: src/lightning/pytorch/loops/training_epoch_loop.py [hunk 2]
```diff
@@ -325,30 +325,33 @@ def advance(self, data_fetcher: _DataFetcher) -> None:
...
-            with trainer.profiler.profile("run_training_batch"):
-                if trainer.lightning_module.automatic_optimization:
-                    # in automatic optimization, there can only be one optimizer
-                    batch_output = self.automatic_optimization.run(trainer.optimizers[0], batch_idx, kwargs)
-                else:
-                    batch_output = self.manual_optimization.run(kwargs)
+            should_skip_rest_of_epoch = response == -1
+            # Signal this is the last batch for the current epoch
+            if should_skip_rest_of_epoch:
+                self.batch_progress.increment_by(0, is_last_batch=True)
```
Comment (SkafteNicki) [inline src/lightning/pytorch/loops/training_epoch_loop.py:340]: What is the logic here from changing from 
`self.batch_progress.increment_processed()`
to 
`self.batch_progress.increment_by(0, is_last_batch=True)`
?

Comment (LTMeyer) [reply src/lightning/pytorch/loops/training_epoch_loop.py:340]: > What is the logic here from changing from `self.batch_progress.increment_processed()`
to `self.batch_progress.increment_by(0, is_last_batch=True)`?

`batch_progress.increment_by` is the only method that can set `is_last_batch` to `True`, which is required to trigger the update of lrs in case of `IterableDataset`. 
The `increment_processed` only increments the counters. In case of `IterableDataset`, for which the expected number of batches is not known, this may not be enough to detect the epoch has ended.
Indeed, the lrs are later updated only if [`num_ready_batches_reached`](https://github.com/Lightning-AI/pytorch-lightning/blob/60a7cd3ef457974a06d420251e2f01775f811bd8/src/lightning/pytorch/loops/training_epoch_loop.py#L361-L362) returns `True`. [It does return `True`](https://github.com/Lightning-AI/pytorch-lightning/blob/60a7cd3ef457974a06d420251e2f01775f811bd8/src/lightning/pytorch/loops/training_epoch_loop.py#L442-L445) if `epoch_finished_on_ready` or `is_last_batch` are `True`. 

> Could we please also add some testing that the changed logic then correctly updates the learning rate when response=-1?

I'll do it.

**Comment by LTMeyer** [comment]
> Could we please also add some testing that the changed logic then correctly updates the learning rate when response=-1?

@SkafteNicki I've just added tests. Not sure about the most appropriate location though. 
https://github.com/Lightning-AI/pytorch-lightning/pull/21307/commits/10d18d91c4c750a180f7be4b10b566e2e2ec097c actually added tests:
- to check `last_batch` is set when rest of epoch was skipped.
- to check lr is being updated at the end of epoch when `on_train_batch_start` returns -1.

I've checked that these tests were indeed failing before the changes introduced in the current PR.

File: src/lightning/fabric/CHANGELOG.md [hunk 3]
```diff
@@ -25,7 +25,11 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
--
+- Fixed `EADDRINUSE` errors in distributed tests with port manager and retry logic ([#21309](https://github.com/Lightning-AI/pytorch-lightning/pull/21309))
```
Comment (deependujha) [inline src/lightning/fabric/CHANGELOG.md:28]: why is this in the changelog for this PR?

Comment (LTMeyer) [reply src/lightning/fabric/CHANGELOG.md:28]: This is the result of a8d9f8bf4f5fd2f122273b6b04dfd114b971e13d, which merged master onto the branch. But it should certainly not have been appeared here.
